### PR TITLE
[Inference] feat rmsnorm cuda kernel and add unittest, benchmark script

### PR DESCRIPTION
**Pull request by SunflowerAries**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

model benchmark:
| bsz | in_len | out_len | Throughput (tokens/sec) |
|-----|--------|---------|--------------------------|
| 16 | 128 | 128 | 1647.18 -> 1756.82 |
| 32 | 128 | 128 | 3000.20 -> 3058.36 |
| 64 | 128 | 128 | 4954.12 -> 4918.43 |
| 16 | 512 | 128 | 1379.97 -> 1353.25 |
| 32 | 512 | 128 | 2053.62 -> 2036.26 |
| 64 | 512 | 128 | 2689.92 -> 2680.61 |
| 16 | 1024 | 128 | 1024.90 -> 1021.70 |
| 32 | 1024 | 128 | 1347.42 -> 1352.80 |
| 64 | 1024 | 128 | 1610.99 -> 1616.78 |

image
below are rmsnorm under hidden size = 1024
RMSNorm benchmarking results
below are rmsnorm under hidden size = 4096
RMSNorm benchmarking results-4095

### CODE REVIEW

File: extensions/inference/inference_ops_cuda.py [hunk 1]
```diff
@@ -12,12 +12,13 @@ def sources_files(self):
...
                 "cuda/colossal_inference_C_frontend.cpp",
                 "cuda/decode_kv_cache_memcpy_kernel.cu",
+                "cuda/rms_layernorm_kernel.cu",
             ]
         ]
         return ret
 
     def include_dirs(self):
-        ret = [self.get_cuda_home_include()]
+        ret = [self.csrc_abs_path("cuda/include"), self.get_cuda_home_include()]
```
Comment (yuehuayingxueluo) [inline extensions/inference/inference_ops_cuda.py:21]: Why is this modification made here?

Comment (SunflowerAries) [reply extensions/inference/inference_ops_cuda.py:21]: The rmsnorm cuda kernel depends on the block_reduce library under cuda/include, thus we need add the library path to the include directories when building the executable

**Review by yuehuayingxueluo — APPROVED** [review]

File: tests/test_infer/test_inference_engine.py [hunk 2]
```diff
@@ -44,7 +40,7 @@ def check_inference_engine(use_engine=False, prompt_template=None):
     top_k = 50
 
     if use_engine:
-        inference_config = InferenceConfig(max_output_len=output_len, prompt_template=prompt_template)
+        inference_config = InferenceConfig(max_output_len=output_len, prompt_template=prompt_template, dtype="fp32")
```
Comment (FrankLeeeee) [inline tests/test_infer/test_inference_engine.py:43]: Why update this to "fp32"?

**Review by FrankLeeeee — APPROVED** [review]
LGTM.

File: colossalai/inference/modeling/models/nopadding_llama.py [hunk 3]
```diff
@@ -185,12 +187,32 @@ def llama_decoder_layer_forward(
     )
 
     # Fully Connected
-    hidden_states, residual = self.post_attention_layernorm(hidden_states, norm_output, residual)
+    hidden_states, residual = self.post_attention_layernorm(hidden_states, norm_output, residual, use_cuda_kernel)
```
Comment (CjhHa1) [inline colossalai/inference/modeling/models/nopadding_llama.py:190]: Do this mean that user can not define which kernel to use?

File: tests/test_infer/test_inference_engine.py (hunk 2 above)
Comment (SunflowerAries) [reply tests/test_infer/test_inference_engine.py:43]: after switching to cuda kernel implementation, now we can not pass the precision test for whole model inference under fp16, thus we need switch to fp32, here dtype will determine the exact data type of model parameters and need to be set to fp32

Comment (SunflowerAries) [reply tests/test_infer/test_inference_engine.py:43]: after switching to cuda kernel implementation, now we can not pass the precision test for whole model inference under fp16, thus we need switch to fp32, here dtype will determine the exact data type of model parameters and need to be set to fp32

**Pull request merged**